### PR TITLE
[SYCL][L0][Plugin] Call ZeCommandQueueCreate on demand

### PR DESCRIPTION
**Pull request by asudarsa**
Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -2803,39 +2865,18 @@ pi_result piQueueCreate(pi_context Context, pi_device Device,
 
   std::vector<ze_command_queue_handle_t> ZeCopyCommandQueues;
 
-  // Create queue to main copy engine
+  // Create 'placeholder queue' to main copy engine
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2868]: please clarify the notion of "placeholder queue"

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -1260,6 +1263,59 @@ bool _pi_queue::isBatchingAllowed() {
...
+  // Return the Ze copy command queue, if already available
+  if (ZeCopyCommandQueues[Index]) {
+    ZeCopyCommandQueue = ZeCopyCommandQueues[Index];
+    return PI_SUCCESS;
+  }
+
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+
+  // Ze copy command queue is not avialable at 'Index'. So we create it.
+  if (Index == 0) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1283]: Why is this the correct check for no queue at "Index"?

Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1283]: I meant that the remaining code in this function is used to create a queue.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -1260,6 +1263,60 @@ bool _pi_queue::isBatchingAllowed() {
...
+            Device->ZeLinkCopyQueueGroupIndex, LinkIndex);
+    ZeCommandQueueDesc.ordinal = Device->ZeLinkCopyQueueGroupIndex;
+    ZeCommandQueueDesc.index = LinkIndex;
+    ZE_CALL(zeCommandQueueCreate,
+            (Context->ZeContext, Device->ZeDevice,
+             &ZeCommandQueueDesc, // TODO: translate properties
+             &ZeCopyCommandQueue));
+    ZeCopyCommandQueues[Index] = ZeCopyCommandQueue;
+    return PI_SUCCESS;
+  }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1316]: the if/else above exepctedly do very similar things, can you combine them please?

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
please address comments

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -1260,6 +1263,60 @@ bool _pi_queue::isBatchingAllowed() {
...
+pi_result _pi_queue::getOrCreateCopyCommandQueue(
+    int Index, ze_command_queue_handle_t &ZeCopyCommandQueue) {
+  ZeCopyCommandQueue = nullptr;
+
+  // Make sure 'Index' is within limits
+  PI_ASSERT((Index >= 0) && (Index < (int)(ZeCopyCommandQueues.size())),
+            PI_INVALID_VALUE);
+
+  // Return the Ze copy command queue, if already available
+  if (ZeCopyCommandQueues[Index]) {
```
[the remark this answers is not in the document]
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1275]: It is not a map. It is a vector. Each entry in the vector will either be a nullptr or an handle to a L0 command queue.

[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1275]: Ah, I missed that. OK, then

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -1260,6 +1263,42 @@ bool _pi_queue::isBatchingAllowed() {
...
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+  ZeCommandQueueDesc.ordinal = (Index == 0) ? Device->ZeMainCopyQueueGroupIndex
+                                            : Device->ZeLinkCopyQueueGroupIndex;
+  // There are two copy queues: main copy queues and link copy queues.
+  // ZeCommandQueueDesc.index is the index into the list of main (or link)
+  // copy queues.
+  // (Index == 0) means we are using the main copy queue.
+  // Otherwise, we use one of the link copy queues.
+  ZeCommandQueueDesc.index =
+      (Index == 0) ? 0 : Index - Device->hasMainCopyEngine();
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1290]:
```suggestion
  ZeCommandQueueDesc.index = Index;
```
Why do you need to subtract anything? (even if needed you should not subtract a boolean)

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1290]: ZeCommandQueueDesc.index is the index into the list of Link Copy Engines (0-7). 'Index' is the index into the list of ALL copy engines (including main copy engine, if present) and it goes from 0-8 (or 0-7, if main copy engine is not available). So, we need to subtract 1 from Index (if main copy engine is present) to get ZeCommandQueueDesc.index for link copy engines.

**Comment by asudarsa** [comment]
I am not sure what's going on with the testing. All tests passed locally for me. And the logs do not show any fails. How can I restart this? Thanks.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1290]: Is it really possible that there are "link" engines, but no "main" engine? If not, then ther will always be either 1 (0 index) engine or multiple (0 index being main, and >0 being link). Anyways, please do not treat a boolean "true" value as "1", as it is only "non-0" by standard.

Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I replaced this with a hard '-1' for link copy indices and modified the comment. Thanks

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Thanks

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Pull request merged**